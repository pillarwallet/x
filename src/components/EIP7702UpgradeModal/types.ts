export type UpgradeStatus =
  | 'ready'
  | 'submitted'
  | 'upgrading'
  | 'completed'
  | 'failed';

export type UserOpUpgradeStatus =
  | 'New'
  | 'Pending'
  | 'Submitted'
  | 'OnChain'
  | 'Finalized'
  | 'Cancelled'
  | 'Reverted';
