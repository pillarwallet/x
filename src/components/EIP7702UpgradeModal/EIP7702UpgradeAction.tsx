import { isNaN } from 'lodash';
import { TailSpin } from 'react-loader-spinner';

// images
import logoEthereum from '../../assets/images/logo-ethereum.png';
import UpgradeGreenLogo from '../../assets/images/upgrade-green-logo.svg';

// utils
import { limitDigitsNumber } from '../../utils/number';

interface EIP7702UpgradeActionProps {
  onClose: () => void;
  handleUpgrade: () => void;
  gasFeeEstimates?: string;
  isCheckingGas?: boolean;
  hasEnoughGas?: boolean;
}

const CheckIcon = () => {
  return (
    <div className="flex items-center justify-center w-4 h-4 rounded-full bg-[#8A77FF] border-[1.5px] border-[#A594FF]">
      <svg
        className="w-3 h-3 text-white"
        fill="currentColor"
        viewBox="0 0 20 20"
        strokeWidth="2"
      >
        <path
          fillRule="evenodd"
          d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
          clipRule="evenodd"
        />
      </svg>
    </div>
  );
};

const EIP7702UpgradeAction = ({
  onClose,
  gasFeeEstimates,
  handleUpgrade,
  isCheckingGas = false,
  hasEnoughGas = false,
}: EIP7702UpgradeActionProps) => {
  return (
    <>
      <div className="flex flex-col gap-2.5 items-center">
        <img
          src={UpgradeGreenLogo}
          alt="upgrade-green-arrows"
          className="w-[66px] h-[44px]"
        />
        <p className="text-white font-normal text-xl">Upgrade your account</p>
        <div className="flex justify-center items-center text-sm font-normal gap-1">
          <p className="text-white">Network:</p>
          <img
            src={logoEthereum}
            className="rounded-full w-4 h-4 flex-shrink-0"
            alt="logo-ethereum"
          />
          <p className="text-[#627EEA]">Ethereum</p>
        </div>
      </div>
      <div className="flex flex-col gap-2">
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Faster Transactions.{' '}
            </span>
            Lower fees.
          </p>
        </div>
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Pay with any token.{' '}
            </span>
            Any time.
          </p>
        </div>
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Same account.{' '}
            </span>
            Smart Features.
          </p>
        </div>
      </div>
      <div className="flex flex-col gap-2 w-full items-center">
        <div className="flex gap-2 items-center">
          <p
            className="text-[13px] text-white font-normal"
            style={{ opacity: 0.7 }}
          >
            Fees:
          </p>
          {isCheckingGas ? (
            <TailSpin color="#FFFFFF" height={12} width={12} />
          ) : (
            <p className="text-[13px] text-white font-normal">
              {gasFeeEstimates && !isNaN(Number(gasFeeEstimates))
                ? limitDigitsNumber(Number(gasFeeEstimates))
                : '-'}{' '}
              ETH
            </p>
          )}
        </div>
        <div className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]">
          <button
            className={`flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] text-white font-normal text-base transition-opacity ${
              isCheckingGas || !hasEnoughGas
                ? 'bg-[#3D3A44] cursor-not-allowed opacity-60'
                : 'bg-[#8A77FF] hover:opacity-90'
            }`}
            type="button"
            onClick={handleUpgrade}
            disabled={isCheckingGas || !hasEnoughGas}
          >
            {!hasEnoughGas && !isCheckingGas ? 'Not enough gas' : 'Upgrade now'}
          </button>
        </div>
        <button
          type="button"
          className="text-white text-xs font-normal underline underline-offset-2 opacity-70"
          onClick={onClose}
        >
          Upgrade later
        </button>
      </div>
    </>
  );
};

export default EIP7702UpgradeAction;
