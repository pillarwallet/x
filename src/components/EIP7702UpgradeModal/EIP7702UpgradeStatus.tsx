import React from 'react';

// types
import { UpgradeStatus } from './types';

// utils
import { getStatusConfig } from '../../utils/upgrade';

// components
import EIP7702UpgradeStatusButton from './EIP7702UpgradeStatusButton';
import EIP7702UpgradeStatusIcon from './EIP7702UpgradeStatusIcon';

interface EIP7702UpgradeStatusProps {
  status: UpgradeStatus;
  onViewDetails: () => void;
}

const EIP7702UpgradeStatus: React.FC<EIP7702UpgradeStatusProps> = ({
  status,
  onViewDetails,
}) => {
  const statusConfig = getStatusConfig(status);

  return (
    <div
      className="flex flex-col items-center justify-center h-full"
      data-testid="pulse-transaction-status-container"
    >
      <EIP7702UpgradeStatusIcon status={status} />
      <div
        className="text-white font-normal text-[20px] mt-3"
        data-testid="pulse-transaction-status-text"
      >
        {statusConfig.label}
      </div>
      <EIP7702UpgradeStatusButton status={status} onClick={onViewDetails} />
    </div>
  );
};

export default React.memo(EIP7702UpgradeStatus);
