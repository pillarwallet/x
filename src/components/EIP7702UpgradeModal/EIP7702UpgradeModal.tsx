import React, { useCallback, useEffect, useRef, useState } from 'react';

// hooks
import { useAppDispatch } from '../../apps/pillarx-app/hooks/useReducerHooks';
import { useEIP7702Upgrade } from '../../hooks/useEIP7702Upgrade';
import { useTransactionDebugLogger } from '../../hooks/useTransactionDebugLogger';
import useTransactionKit from '../../hooks/useTransactionKit';

// reducer
import {
  setHasCompletedEIP7702Upgrade,
  setIsEIP7702Eligible,
} from '../../apps/pillarx-app/reducer/WalletPortfolioSlice';

// services
import { getUserOperationStatus } from '../../services/userOpStatus';

// types
import { UpgradeStatus, UserOpUpgradeStatus } from './types';

// components
import EIP7702UpgradeAction from './EIP7702UpgradeAction';
import EIP7702UpgradeDetails from './EIP7702UpgradeDetails';
import EIP7702UpgradeStatus from './EIP7702UpgradeStatus';

interface EIP7702UpgradeModalProps {
  isOpen: boolean;
  onClose: () => void;
}

const EIP7702UpgradeModal: React.FC<EIP7702UpgradeModalProps> = ({
  isOpen,
  onClose,
}) => {
  const dispatch = useAppDispatch();
  const { kit } = useTransactionKit();
  const { transactionDebugLog } = useTransactionDebugLogger();
  const [upgradeStatus, setUpgradeStatus] = useState<UpgradeStatus>('ready');
  const { calculateGasFees, setGasUpgradeInfo, gasUpgradeInfo, isCheckingGas } =
    useEIP7702Upgrade();
  const [userOpHash, setUserOpHash] = useState<string>('');
  const [isPolling, setIsPolling] = useState(false);
  const [errorMessage, setErrorMessage] = useState<string>('');
  const [statusStartTime, setStatusStartTime] = useState<number>(Date.now());
  const [hasSeenSuccess, setHasSeenSuccess] = useState<boolean>(false);
  const [failureTimeoutId, setFailureTimeoutId] = useState<ReturnType<
    typeof setTimeout
  > | null>(null);
  const [showDetails, setShowDetails] = useState<boolean>(false);
  const [submittedAt, setSubmittedAt] = useState<Date | undefined>(undefined);
  const [pendingCompletedAt, setPendingCompletedAt] = useState<
    Date | undefined
  >(undefined);
  const hasSeenSuccessRef = useRef<boolean>(false);
  const blockchainTxHashRef = useRef<string | undefined>(undefined);
  const failureGraceExpiryRef = useRef<number | null>(null);
  const hasCalculatedGasThisSessionRef = useRef<boolean>(false);

  // Handle close with localStorage persistence
  const handleClose = useCallback(() => {
    // Mark that user dismissed the upgrade modal
    localStorage.setItem('eip7702_upgrade_dismissed', 'true');
    onClose();
  }, [onClose]);

  // Map UserOp status to UpgradeStatus
  const mapUserOpStatusToUpgradeStatus = (
    status: UserOpUpgradeStatus
  ): UpgradeStatus => {
    if (status === 'New') {
      return 'submitted';
    }
    if (status === 'Pending' || status === 'Submitted') {
      return 'upgrading';
    }
    if (status === 'OnChain' || status === 'Finalized') {
      return 'completed';
    }
    if (status === 'Cancelled' || status === 'Reverted') {
      return 'failed';
    }
    return 'submitted';
  };

  // Function to update status with minimum display duration
  const updateStatusWithDelay = useCallback(
    (newStatus: UpgradeStatus) => {
      const now = Date.now();
      const timeSinceLastStatus = now - statusStartTime;
      const minDisplayTime = 2000; // 2 seconds
      const failureConfirmationTime = 10000; // 10 seconds for failed status

      // If we're already on a final status, stop polling
      const isFinalStatus =
        upgradeStatus === 'completed' || upgradeStatus === 'failed';
      if (isFinalStatus) {
        setIsPolling(false);
        return;
      }

      // If we get a success status, cancel any pending failure timeout
      if (newStatus === 'completed') {
        if (failureTimeoutId) {
          clearTimeout(failureTimeoutId);
          setFailureTimeoutId(null);
        }
        setHasSeenSuccess(true);
        hasSeenSuccessRef.current = true;
        // If we're currently showing pending due to a failure confirmation,
        // immediately show the success status
        if (upgradeStatus === 'upgrading') {
          setUpgradeStatus('completed');
          setStatusStartTime(now);
          setIsPolling(false);
          return;
        }
      }

      // Special handling for failed status - wait 10 seconds before confirming
      if (newStatus === ('failed' as UpgradeStatus)) {
        // If we've already seen success, ignore failure statuses
        if (hasSeenSuccess || hasSeenSuccessRef.current) {
          return;
        }

        // Start the confirmation timer for failed status
        setUpgradeStatus('upgrading'); // Keep showing upgrading while we wait
        setStatusStartTime(now);

        // After 10 seconds, confirm the failure
        const timeoutId = setTimeout(() => {
          // Double-check that we haven't seen success in the meantime
          if (hasSeenSuccessRef.current) {
            setFailureTimeoutId(null);
            return;
          }
          setUpgradeStatus('failed');
          setStatusStartTime(Date.now());
          setIsPolling(false); // Stop polling after confirming failure
          setFailureTimeoutId(null);
        }, failureConfirmationTime);
        setFailureTimeoutId(timeoutId);
        return;
      }

      // If this is a final status and we haven't shown intermediate states yet,
      // we need to show them first
      if (
        (newStatus === 'completed' || newStatus === 'failed') &&
        (upgradeStatus === 'submitted' || upgradeStatus === 'upgrading')
      ) {
        // Show intermediate states first
        if (upgradeStatus === 'submitted') {
          setUpgradeStatus('upgrading');
          setStatusStartTime(now);

          // After 2 seconds, show the final status
          setTimeout(() => {
            setUpgradeStatus(newStatus);
            setStatusStartTime(Date.now());
            setIsPolling(false); // Stop polling after showing final status
          }, minDisplayTime);
          return;
        }
      }

      if (timeSinceLastStatus < minDisplayTime) {
        // If not enough time has passed, delay the status update
        const remainingTime = minDisplayTime - timeSinceLastStatus;
        setTimeout(() => {
          setUpgradeStatus(newStatus);
          setStatusStartTime(Date.now());
        }, remainingTime);
      } else {
        // Enough time has passed, update immediately
        setUpgradeStatus(newStatus);
        setStatusStartTime(now);

        // Stop polling for final statuses
        if (
          newStatus === 'completed' ||
          newStatus === ('failed' as UpgradeStatus)
        ) {
          setIsPolling(false);
        }
      }
    },
    [upgradeStatus, statusStartTime, failureTimeoutId, hasSeenSuccess]
  );

  const handleUpgrade = async () => {
    // Reset state
    setUserOpHash('');
    setErrorMessage('');
    setHasSeenSuccess(false);
    hasSeenSuccessRef.current = false;
    blockchainTxHashRef.current = undefined;
    failureGraceExpiryRef.current = null;
    setStatusStartTime(Date.now());
    setUpgradeStatus('submitted');

    transactionDebugLog('Initiating EIP-7702 upgrade on chain 1 (Mainnet)');

    try {
      if (!kit) {
        console.error('Transaction kit not available');
        transactionDebugLog('Upgrade failed - transaction kit not available');
        setUpgradeStatus('failed');
        setErrorMessage(
          'Oops something went wrong. Please refresh the page and try again.'
        );
        return;
      }

      // Call delegateSmartAccountToEoa to initiate upgrade
      const result = await kit.delegateSmartAccountToEoa({
        chainId: 1,
        delegateImmediately: true,
      });

      transactionDebugLog(
        'EIP-7702 upgrade delegateSmartAccountToEoa result:',
        {
          userOpHash: result.userOpHash,
          isAlreadyInstalled: result.isAlreadyInstalled,
        }
      );

      if (result.userOpHash && result.userOpHash.trim() !== '') {
        transactionDebugLog(
          'EIP-7702 upgrade initiated successfully with userOpHash:',
          result.userOpHash
        );
        setUserOpHash(result.userOpHash);
        setIsPolling(true);
      } else if (result.isAlreadyInstalled) {
        transactionDebugLog('EIP-7702 already installed');
        updateStatusWithDelay('completed');
      } else {
        transactionDebugLog('EIP-7702 authorization signed but not executed');
        updateStatusWithDelay('completed');
      }
    } catch (error) {
      console.error('EIP-7702 upgrade failed:', error);
      transactionDebugLog('EIP-7702 upgrade failed with error:', error);
      setUpgradeStatus('failed');
      setErrorMessage(
        'The upgrade failed. Please try again later or contact support.'
      );
    }
  };

  // Polling effect for status updates
  useEffect(() => {
    const chainId = 1;
    if (!userOpHash || !chainId || !isPolling) {
      return;
    }

    let pollAttempt = 0;
    const pollStatus = async () => {
      // Check if polling is still active before making the API call
      if (!isPolling) {
        return;
      }

      pollAttempt += 1;
      transactionDebugLog(
        `Polling EIP-7702 upgrade status (attempt ${pollAttempt}) for userOpHash:`,
        userOpHash
      );

      try {
        const response = await getUserOperationStatus(chainId, userOpHash);
        transactionDebugLog(
          `UserOp status response (attempt ${pollAttempt}):`,
          {
            status: response?.status,
            transaction: response?.transaction,
            reason: response?.reason,
            error: response?.error,
          }
        );

        // Handle case where response is undefined/null (API error)
        if (!response) {
          transactionDebugLog('No response from getUserOperationStatus');
          return;
        }

        if (response.status) {
          const newUserOpStatus = response.status as UserOpUpgradeStatus;
          const newUpgradeStatus =
            mapUserOpStatusToUpgradeStatus(newUserOpStatus);

          // Track if we've seen success - once we have, ignore any failure statuses
          if (newUpgradeStatus === 'completed') {
            transactionDebugLog(
              'EIP-7702 upgrade status changed to completed',
              {
                userOpStatus: newUserOpStatus,
                transactionHash: response.transaction,
              }
            );
            setHasSeenSuccess(true);
            hasSeenSuccessRef.current = true;
          }

          // If we've already seen success, ignore any failure statuses
          // Both state and ref to ensure we catch the success
          if (
            (hasSeenSuccess || hasSeenSuccessRef.current) &&
            newUpgradeStatus === 'failed'
          ) {
            return;
          }

          // Extract actual transaction hash if available and mirror into ref (do this before failure guard)
          if (response.transaction) {
            const firstObservation = !blockchainTxHashRef.current;
            blockchainTxHashRef.current = response.transaction;
            if (firstObservation) {
              transactionDebugLog(
                'Blockchain transaction hash observed for EIP-7702 upgrade:',
                response.transaction
              );
              if (failureGraceExpiryRef.current == null) {
                failureGraceExpiryRef.current = Date.now() + 10000; // start grace on first hash
              }
            }
          }

          // Additional protection using refs and single-use grace window
          // If we have a blockchain transaction hash, it means the transaction was actually
          // submitted and might be successful even if the status shows as 'Reverted' temporarily
          if (newUpgradeStatus === 'failed') {
            const nowTs = Date.now();
            const hasOnChainHash = Boolean(blockchainTxHashRef.current);
            if (hasOnChainHash) {
              // Initialize grace window if not set
              if (failureGraceExpiryRef.current == null) {
                failureGraceExpiryRef.current = nowTs + 10000; // 10s grace
                return; // skip once immediately when hash first observed
              }
              if (nowTs < failureGraceExpiryRef.current) {
                return; // still within grace window
              }
            }
          }

          // Always use delayed update to ensure all states are shown
          transactionDebugLog(
            `EIP-7702 upgrade status update: ${newUpgradeStatus} (from UserOp status: ${newUserOpStatus})`
          );
          updateStatusWithDelay(newUpgradeStatus);

          // Capture error details if transaction failed
          if (newUpgradeStatus === 'failed') {
            transactionDebugLog('EIP-7702 upgrade failed', {
              userOpStatus: newUserOpStatus,
              transactionHash: response.transaction,
              reason: response.reason,
              error: response.error,
            });
            setErrorMessage('Upgrade failed. Please try again.');
          }
        }
      } catch (error) {
        console.error('Failed to get transaction status:', error);
        transactionDebugLog('Error getting EIP-7702 upgrade status:', error);
        updateStatusWithDelay('failed');
        setErrorMessage('Failed to get upgrade status');
      }
    };

    // Poll immediately
    pollStatus();

    // Set up polling every 2 seconds
    const interval = setInterval(pollStatus, 2000);

    // eslint-disable-next-line consistent-return
    return () => {
      clearInterval(interval);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userOpHash, isPolling, hasSeenSuccess, updateStatusWithDelay]);

  // When upgrade completes successfully, mark as completed and not eligible
  useEffect(() => {
    if (upgradeStatus === 'completed') {
      transactionDebugLog('EIP-7702 upgrade completed successfully', {
        userOpHash,
        transactionHash: blockchainTxHashRef.current,
      });
      // Mark that upgrade has been completed successfully
      dispatch(setHasCompletedEIP7702Upgrade(true));
      // Mark as not eligible to hide the upgrade button
      dispatch(setIsEIP7702Eligible(false));
      // Clear dismissal flag since upgrade is now complete
      localStorage.removeItem('eip7702_upgrade_dismissed');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [upgradeStatus, dispatch, userOpHash]);

  // Track when steps complete - this runs every time status changes
  useEffect(() => {
    const now = new Date();

    // Submitted step is completed when we move from 'submitted' to 'upgrading' or beyond
    // Set the timestamp when we first see the step as completed
    if (
      upgradeStatus === 'upgrading' ||
      upgradeStatus === 'completed' ||
      upgradeStatus === 'failed'
    ) {
      // Only set if not already set (first time this status is reached)
      setSubmittedAt((prev) => prev || now);
    }

    // Upgrading step is completed when status is 'completed' or 'failed'
    if (upgradeStatus === 'completed' || upgradeStatus === 'failed') {
      // Only set if not already set (first time this status is reached)
      setPendingCompletedAt((prev) => prev || now);
    }
  }, [upgradeStatus]);

  // Calculate gas when modal opens (once per open session)
  useEffect(() => {
    if (!isOpen) {
      hasCalculatedGasThisSessionRef.current = false;
      return;
    }

    if (!calculateGasFees || !setGasUpgradeInfo) {
      return;
    }

    if (hasCalculatedGasThisSessionRef.current) {
      return;
    }

    hasCalculatedGasThisSessionRef.current = true;

    const fetchGas = async () => {
      try {
        transactionDebugLog('Starting gas calculation for EIP-7702 upgrade');
        const result = await calculateGasFees();
        if (result && setGasUpgradeInfo) {
          transactionDebugLog('Gas calculation completed:', {
            balanceInEth: result.balanceInEth,
            gasCostInEth: result.gasCostInEth,
            requiredEth: result.requiredEth,
            hasEnoughEth: result.hasEnoughEth,
          });
          setGasUpgradeInfo(result);
        } else {
          transactionDebugLog('Gas calculation returned null');
        }
      } catch (error) {
        console.error('Failed to calculate gas for EIP-7702 upgrade:', error);
        transactionDebugLog('Gas calculation failed:', error);
        // Set a default state so user can still see the modal
        if (setGasUpgradeInfo) {
          setGasUpgradeInfo({
            balanceInEth: 0,
            gasCostInEth: 0,
            requiredEth: 0,
            hasEnoughEth: false,
          });
        }
      }
    };

    fetchGas();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isOpen, calculateGasFees, setGasUpgradeInfo]);

  // Render different components based on status
  const renderContent = () => {
    // If user is viewing details
    if (showDetails) {
      return (
        <EIP7702UpgradeDetails
          onClose={() => setShowDetails(false)}
          onCloseModal={onClose}
          status={upgradeStatus}
          submittedAt={submittedAt}
          pendingCompletedAt={pendingCompletedAt}
          errorDetails={errorMessage}
        />
      );
    }

    // Initial state - show action/benefits
    if (upgradeStatus === 'ready') {
      return (
        <EIP7702UpgradeAction
          onClose={handleClose}
          gasFeeEstimates={gasUpgradeInfo?.requiredEth.toString()}
          handleUpgrade={handleUpgrade}
          isCheckingGas={isCheckingGas}
          hasEnoughGas={gasUpgradeInfo?.hasEnoughEth ?? false}
        />
      );
    }

    // Status view (submitted, upgrading, completed, failed)
    return (
      <EIP7702UpgradeStatus
        status={upgradeStatus}
        onViewDetails={() => setShowDetails(true)}
      />
    );
  };

  // Early return after all hooks have been called
  if (!isOpen) return null;

  // Determine padding based on what's being shown
  const shouldShowLargePadding = !showDetails && upgradeStatus !== 'ready';

  return (
    <div
      className="fixed inset-0 flex items-center justify-center pt-4 pb-20 z-50"
      style={{
        backgroundColor: 'rgba(18, 17, 22, 0.8)',
        backdropFilter: 'blur(10px)',
      }}
    >
      <div
        className={`flex flex-col bg-[#1E1D24] rounded-2xl px-3 max-w-md mx-4 w-full items-center gap-9 max-h-[calc(100vh-6rem)] overflow-y-auto ${shouldShowLargePadding ? 'py-20' : 'py-6'}`}
      >
        {renderContent()}
      </div>
    </div>
  );
};

export default EIP7702UpgradeModal;
