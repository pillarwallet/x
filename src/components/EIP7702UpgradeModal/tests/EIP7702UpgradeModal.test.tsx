/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// components
import EIP7702UpgradeModal from '../EIP7702UpgradeModal';

// Mock dependencies
vi.mock('../../../hooks/useEIP7702Upgrade', () => ({
  useEIP7702Upgrade: vi.fn(() => ({
    calculateGasFees: vi.fn().mockResolvedValue({
      balanceInEth: 0.1,
      gasCostInEth: 0.001,
      requiredEth: 0.002,
      hasEnoughEth: true,
    }),
    setGasUpgradeInfo: vi.fn(),
    gasUpgradeInfo: {
      balanceInEth: 0.1,
      gasCostInEth: 0.001,
      requiredEth: 0.002,
      hasEnoughEth: true,
    },
    isCheckingGas: false,
    isEligible: true,
    handleUpgradeClick: vi.fn(),
    checkOnLogin: vi.fn(),
  })),
}));

vi.mock('../../../hooks/useTransactionKit', () => ({
  default: vi.fn(() => ({
    kit: {
      getEtherspotProvider: vi.fn(() => ({
        getWalletMode: vi.fn(() => 'modular'),
      })),
      delegateSmartAccountToEoa: vi.fn().mockResolvedValue({
        userOpHash: '0xTestHash123',
        isAlreadyInstalled: false,
      }),
    },
  })),
}));

vi.mock('../../../hooks/useTransactionDebugLogger', () => ({
  useTransactionDebugLogger: vi.fn(() => ({
    transactionDebugLog: vi.fn(),
  })),
}));

vi.mock('../../../services/userOpStatus', () => ({
  getUserOperationStatus: vi.fn().mockResolvedValue({
    status: 'OnChain',
    transaction: '0xTransactionHash',
  }),
}));

vi.mock('../../../apps/pillarx-app/reducer/WalletPortfolioSlice', () => ({
  setIsUpgradeWalletModalOpen: vi.fn(() => ({
    type: 'setIsUpgradeWalletModalOpen',
  })),
  setIsEIP7702Eligible: vi.fn(() => ({ type: 'setIsEIP7702Eligible' })),
  setHasCompletedEIP7702Upgrade: vi.fn(() => ({
    type: 'setHasCompletedEIP7702Upgrade',
  })),
}));

vi.mock('../../../apps/pillarx-app/hooks/useReducerHooks', () => ({
  useAppDispatch: vi.fn(() => vi.fn()),
}));

// Mock child components
vi.mock('../EIP7702UpgradeAction', () => ({
  default: ({
    onClose,
    handleUpgrade,
    hasEnoughGas,
    isCheckingGas,
  }: {
    onClose: () => void;
    handleUpgrade: () => void;
    hasEnoughGas: boolean;
    isCheckingGas: boolean;
  }) => (
    <div data-testid="upgrade-action">
      <button
        type="button"
        data-testid="upgrade-button"
        onClick={handleUpgrade}
        disabled={!hasEnoughGas || isCheckingGas}
      >
        Upgrade
      </button>
      <button type="button" data-testid="close-button" onClick={onClose}>
        Close
      </button>
    </div>
  ),
}));

vi.mock('../EIP7702UpgradeStatus', () => ({
  default: ({
    status,
    onViewDetails,
  }: {
    status: string;
    onViewDetails: () => void;
  }) => (
    <div data-testid={`upgrade-status-${status}`}>
      <button type="button" data-testid="view-details" onClick={onViewDetails}>
        View Details
      </button>
    </div>
  ),
}));

vi.mock('../EIP7702UpgradeDetails', () => ({
  default: ({ onClose, status }: { onClose: () => void; status: string }) => (
    <div data-testid={`upgrade-details-${status}`}>
      <button type="button" data-testid="close-details" onClick={onClose}>
        Close Details
      </button>
    </div>
  ),
}));

const baseProps = {
  isOpen: true,
  onClose: vi.fn(),
};

describe('<EIP7702UpgradeModal />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    // Clear localStorage before each test
    localStorage.clear();
  });

  describe('Rendering', () => {
    it('renders nothing when isOpen is false', () => {
      render(<EIP7702UpgradeModal {...baseProps} isOpen={false} />);

      expect(screen.queryByTestId('upgrade-action')).not.toBeInTheDocument();
    });

    it('renders correctly and matches snapshot', () => {
      const tree = renderer
        .create(<EIP7702UpgradeModal {...baseProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders action view when status is ready', () => {
      render(<EIP7702UpgradeModal {...baseProps} />);

      expect(screen.getByTestId('upgrade-action')).toBeInTheDocument();
    });
  });

  describe('User interactions - Initial state', () => {
    it('calls onClose and sets localStorage when close is clicked', () => {
      render(<EIP7702UpgradeModal {...baseProps} />);

      const closeButton = screen.getByTestId('close-button');
      fireEvent.click(closeButton);

      expect(baseProps.onClose).toHaveBeenCalled();
      expect(localStorage.getItem('eip7702_upgrade_dismissed')).toBe('true');
    });
  });

  describe('Upgrade flow', () => {
    it('initiates upgrade when upgrade button is clicked', () => {
      render(<EIP7702UpgradeModal {...baseProps} />);

      const upgradeButton = screen.getByTestId('upgrade-button');
      expect(upgradeButton).toBeInTheDocument();
      fireEvent.click(upgradeButton);

      // Verify button was clicked (complex async behavior makes full flow testing difficult)
      expect(true).toBe(true);
    });

    it('disables upgrade button when insufficient gas', async () => {
      const useEIP7702Upgrade = await import(
        '../../../hooks/useEIP7702Upgrade'
      );

      vi.mocked(useEIP7702Upgrade.useEIP7702Upgrade).mockReturnValue({
        calculateGasFees: vi.fn().mockResolvedValue({
          balanceInEth: 0.1,
          gasCostInEth: 0.001,
          requiredEth: 0.002,
          hasEnoughEth: false,
        }),
        setGasUpgradeInfo: vi.fn(),
        gasUpgradeInfo: {
          balanceInEth: 0.1,
          gasCostInEth: 0.001,
          requiredEth: 0.002,
          hasEnoughEth: false,
        },
        isCheckingGas: false,
        isEligible: true,
        handleUpgradeClick: vi.fn(),
        checkOnLogin: vi.fn(),
      });

      render(<EIP7702UpgradeModal {...baseProps} />);

      const upgradeButton = screen.getByTestId('upgrade-button');
      expect(upgradeButton).toBeDisabled();
    });

    it('shows loading when checking gas', async () => {
      const useEIP7702Upgrade = await import(
        '../../../hooks/useEIP7702Upgrade'
      );

      vi.mocked(useEIP7702Upgrade.useEIP7702Upgrade).mockReturnValue({
        calculateGasFees: vi.fn().mockResolvedValue({
          balanceInEth: 0.1,
          gasCostInEth: 0.001,
          requiredEth: 0.002,
          hasEnoughEth: true,
        }),
        setGasUpgradeInfo: vi.fn(),
        gasUpgradeInfo: {
          balanceInEth: 0.1,
          gasCostInEth: 0.001,
          requiredEth: 0.002,
          hasEnoughEth: true,
        },
        isCheckingGas: true,
        isEligible: true,
        handleUpgradeClick: vi.fn(),
        checkOnLogin: vi.fn(),
      });

      render(<EIP7702UpgradeModal {...baseProps} />);

      const upgradeButton = screen.getByTestId('upgrade-button');
      expect(upgradeButton).toBeDisabled();
    });
  });

  describe('Edge cases', () => {
    it('renders action view when modal is opened', () => {
      render(<EIP7702UpgradeModal {...baseProps} />);

      expect(screen.getByTestId('upgrade-action')).toBeInTheDocument();
    });
  });

  describe('LocalStorage behavior', () => {
    it('sets dismissal flag when modal is closed', () => {
      render(<EIP7702UpgradeModal {...baseProps} />);

      const closeButton = screen.getByTestId('close-button');
      fireEvent.click(closeButton);

      expect(localStorage.getItem('eip7702_upgrade_dismissed')).toBe('true');
    });
  });
});
