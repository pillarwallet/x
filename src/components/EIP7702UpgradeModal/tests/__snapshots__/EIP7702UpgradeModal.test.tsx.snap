// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EIP7702UpgradeModal /> > Rendering > renders correctly and matches snapshot 1`] = `
<div
  className="fixed inset-0 flex items-center justify-center pt-4 pb-20 z-50"
  style={
    {
      "backdropFilter": "blur(10px)",
      "backgroundColor": "rgba(18, 17, 22, 0.8)",
    }
  }
>
  <div
    className="flex flex-col bg-[#1E1D24] rounded-2xl px-3 max-w-md mx-4 w-full items-center gap-9 max-h-[calc(100vh-6rem)] overflow-y-auto py-6"
  >
    <div
      data-testid="upgrade-action"
    >
      <button
        data-testid="upgrade-button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Upgrade
      </button>
      <button
        data-testid="close-button"
        onClick={[Function]}
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
